File: 다이내믹 프로그래밍/7579.앱.js
//[골드3] 7579.앱 https://www.acmicpc.net/problem/7579

import { createRequire } from "module";
const require = createRequire(import.meta.url);

const fs = require('fs');
let input = fs.readFileSync('dev/stdin', 'utf-8').trim().split('\n');

let [N, M] = input[0].split(' ').map((V) => +V);
let A = input[1].split(' ').map((v) => +v);
let C = input[2].split(' ').map((v) => +v);

function solution(N, M, A, C) {
    let data = [];
    let sumC = 0;

    for (let i = 0; i < N; i++) {
        data.push([A[i], C[i]]);
        sumC += C[i];
    }

    let dp = Array.from({ length: sumC + 1 }, () => 0);

    for (let i = 0; i < N; i++) {
        let [a, c] = data[i];

        for (let j = sumC; j >= c; j--) {
            dp[j] = Math.max(dp[j], dp[j - c] + a);
        }
    }

    for(let i=0;i<sumC+1;i++){
        if(dp[i] >= M){
            return i;
        }
    }

}

console.log(solution(N, M, A, C));

File: 다이내믹 프로그래밍/7579.앱.cc
//[골드3] 7579.앱 https://www.acmicpc.net/problem/7579

#include <iostream>
#include <algorithm>

using namespace std;

int n, m;
int memory[102];
int cost[102];
int dp[10001][101];
int costSum=0;

void getInput(){
    cin>>n>>m;

    for(int i=1;i<=n;++i){
        cin>>memory[i];
    }
    for(int i=1;i<=n;++i){
        cin>>cost[i];
        costSum+=cost[i];
    }
}

int main(void){

    getInput();

    int answer = -1;

    for(int i=0;i<=costSum;++i){
        for(int j=1;j<=n;++j){
            dp[i][j] = dp[i][j-1];
            if(i - cost[j] >= 0){
                dp[i][j] = max(dp[i][j], dp[i-cost[j]][j-1] + memory[j]);
            }
        }
    }

    // for(int i=0;i<=costSum;++i){
    //     for(int j=0;j<=n;++j){
    //         cout<<dp[i][j]<<" ";
    //     }cout<<endl;
    // }

    for(int i = 0; i <= costSum; ++i){
        if(dp[i][n] >= m){
            answer = i;
            break;
        }
    }

    cout<<answer;

}